packages/spectral/src/types/ConnectionDefinition.ts

```diff
@@ -54,6 +60,65 @@ export interface OnPremConnectionDefinition extends BaseConnectionDefinition {
   };
 }
 
+type TTemplateComplement<TRequired, TGiven> = {
+  [K in Exclude<keyof TRequired, keyof TGiven>]: ConnectionTemplateInputField;
+};
+
+interface OAuth2AuthorizationCodeInputs {
+  /**
+   * The OAuth 2.0 authorization URL where users can consent to granting you permissions
+   * to their third-party account. (e.g. `https://app.acme.com/oauth2/authorize`)
+   */
+  authorizeUrl: ConnectionInput;
+  /**
+   * The OAuth 2.0 token URL which can be used to exchange an auth code or refresh
+   * token for an access token. (e.g. `https://app.acme.com/oauth2/token`)
+   */
+  tokenUrl: ConnectionInput;
+  /** OAuth 2.0 permissions (scopes) to request from the authenticated user. */
+  scopes: ConnectionInput;
+  /** OAuth 2.0 client ID (sometimes called app key or client key). */
+  clientId: ConnectionInput;
+  /** OAuth 2.0 client secret (sometimes called app secret or secret key). */
+  clientSecret: ConnectionInput;
+}
+
+interface OAuth2ClientCredentialInputs {
+  /**
+   * The OAuth 2.0 token URL which can be used to exchange a client ID and secret
+   * for an access token. (e.g. `https://app.acme.com/oauth2/token`)
+   */
+  tokenUrl: ConnectionInput;
+  /** OAuth 2.0 permissions (scopes) to request from the authenticated user. */
+  scopes: ConnectionInput;
+  /** OAuth 2.0 client ID (sometimes called app key or client key). */
+  clientId: ConnectionInput;
+  /** OAuth 2.0 client secret (sometimes called app secret or secret key). */
+  clientSecret: ConnectionInput;
+}
+
+export function templateConnectionInputs<
+  TConnectionType extends "client_credentials" | "authorization_code" | null = null,
+  TInputs extends Record<string, ConnectionInput> = Record<string, ConnectionInput>,
+>(inputSet: {
+  authType?: TConnectionType;
+  inputs: TConnectionType extends null | undefined
+    ? Record<string, ConnectionInput>
+    : TConnectionType extends OAuth2Type.ClientCredentials
+      ? TInputs & Partial<OAuth2ClientCredentialInputs> & { [key: string]: ConnectionInput }
+      : TInputs & Partial<OAuth2AuthorizationCodeInputs> & { [key: string]: ConnectionInput };
+  templateInputs: TConnectionType extends null | undefined
+    ? { [key: string]: ConnectionTemplateInputField }
+    : TConnectionType extends "client_credentials"
+      ? TTemplateComplement<OAuth2ClientCredentialInputs, TInputs> & {
+          [key: string]: ConnectionTemplateInputField;
+        }
+      : TTemplateComplement<OAuth2AuthorizationCodeInputs, TInputs> & {
+          [key: string]: ConnectionTemplateInputField;
+        };
+}) {
+  return merge(inputSet.inputs, inputSet.templateInputs);
+}
```

Comment:
In order to future-proof this against breaking changes we'll want this as multiple arguments instead of an object with `inputs` and `templateInputs` properties. When we do provide stronger typing for the available template tags based on the provided inputs it'll require that we know the `inputs` in order to infer the types for the `templateInputs`. In general, objects don't play as nicely with these kinds of progressive types since there's nothing stopping you from first providing the `templateInputs` property and then moving to `inputs`. In these cases it's better to use multiple params.